Target size feature

gerardo-rodriguez opened the pull request:
Closes #72

gerardo-rodriguez commented on src/user-util/index.ts:
```diff
@@ -45,6 +45,35 @@ ${el}`;
+
+  const display = getComputedStyle(el).display;
+
+  // Per the W3C recommendation, ignore inline elements because they are likely in a sentence
+  if (display === 'inline') {
+    return;
+  }
```
```suggestion
  // Per the W3C recommendation, inline elements are excluded
  // See: https://www.w3.org/WAI/WCAG21/Understanding/target-size.html
  if (display === 'inline') {
    return;
  }
```

gerardo-rodriguez commented on tests/user/click.test.ts:
```diff
@@ -158,13 +156,61 @@ describe('actionability checks', () => {
+      `"Cannot click element that is too small.Target size of element does not meet W3C recommendation of 44px × 44px: https://www.w3.org/WAI/WCAG21/Understanding/target-size.htmlElement was 2px × 2px<button style=\\"width: 2px; height: 2px; border: none; padding: 0;\\">hi</button>You can customize the minimum target size by passing the user.targetSize option to configureDefaults or withBrows ...
+    );
+
+    await user.click(button, { targetSize: 2 });
```
```suggestion
    // This confirms a passing test when setting a custom target size
    await user.click(button, { targetSize: 2 });
```

gerardo-rodriguez commented on tests/user/click.test.ts:
```diff
@@ -158,13 +156,62 @@ describe('actionability checks', () => {
+    await utils.injectHTML(`
+      <p>This is text <a href="#">with a link</a></p>
+    `);
+    // Elements inside of text don't have the 44px × 44px minimum, per W3C recommendation
```
```suggestion
    // Inline elements don't have the 44px × 44px minimum, per W3C recommendation
```

calebeby commented:
@gerardo-rodriguez It looks like the unconventional newline character doesn't get displayed in CI on GitHub. Starting to think maybe we should remove it

calebeby commented:
@gerardo-rodriguez Do you anticipate having time to finish this up? If not, I can.

gerardo-rodriguez commented on src/ansi-colors-browser.ts:
```diff
@@ -16,21 +16,24 @@ export const ansiColorsLog = (...input: unknown[]) => {
-    });
+    str += segment
+      // Convert back from "next-line" character into regular newline for display in browser
+      .replace(/\u0085/g, '\n')
```
> @gerardo-rodriguez It looks like the unconventional newline character doesn't get displayed in CI on GitHub. Starting to think maybe we should remove it

@calebeby Is this the line for the unconventional newline character? I can't remember it's been so long now. 😅

calebeby replied:
Yeah this is one place where it is, I think there is another place too but I can't find it rn because I'm on my phone

gerardo-rodriguez replied:
Thanks! I can look. 🔍 👀

gerardo-rodriguez commented:
I'm not sure if I was supposed to [update the snapshots](https://github.com/cloudfour/pleasantest/pull/248/commits/cb8552868c017294e72fff1912d058757590491e). 🤔

gerardo-rodriguez commented on tests/user/click.test.ts:
```diff
@@ -158,13 +156,62 @@ describe('actionability checks', () => {
+      `"Cannot click element that is too small.Target size of element is smaller than 46px × 46pxElement was 2px × 2px<button style=\\"width: 2px; height: 2px; border: none; padding: 0;\\">hi</button>You can customize the minimum target size by passing the user.targetSize option to configureDefaults or withBrowser or user.click"`,
+    );
+  }),
+);
```
Oh...hmm...good point.

Now the question is, what is the expectation? Are checkboxes exempt? Only exempt when using default styles?

I'm not sure. 🤔

gerardo-rodriguez commented:
@calebeby If you have any time to team up on this to try and knock it out in the coming week, let me know and we can schedule some time. 😉

calebeby commented:
I should have time on Monday, I don't have classes! We can work out a time then

calebeby commented:
Marking this as ready for review, but also once @gerardo-rodriguez is back he will need to review the docs changes I made as well.

calebeby commented:
TODO: Add changeset and decide whether to leave this open instead of merging it right away since it is a breaking change

It may make sense to do a minor release with accessibility snapshots, followed by a major change with the CSS changes I am working on as well as this PR.

Paul-Hebert commented on README.md:
```diff
@@ -324,6 +324,12 @@ Ensures that the element is visible to a user. Currently, the following checks a
+
+Ensures that the clickable region of an element is large enough to easily be activated by users, even users with reduced fine motor skills or while using a touch input device.
+
+Element must be at least 44px wide and at least 44px tall to pass the target size check (configurable via `user.targetSize` option in `WithBrowserOpts` or `targetSize` option for `user.click`).
```
Can we add a link to where these numbers came from? (Is it the Apple HCI guidelines?)

gerardo-rodriguez replied:
We can use this link: https://www.w3.org/TR/WCAG21/#target-size

Paul-Hebert commented on README.md:
```diff
@@ -360,6 +366,8 @@ Call Signatures:
   - `envVars`: Object with string keys and string values for environment variables to pass in as `import.meta.env.*` / `process.env.*`
   - `esbuild`: [`TransformOptions`](https://esbuild.github.io/api/#transform-api) | `false`: Options to pass to esbuild. Set to false to disable esbuild.
+- `user`: User API options object (all properties are optional). They will be applied when calling `user.*` methods.
+  - `targetSize`: (`number | boolean`) Set the minimum target size for `user.click`. Set to `false` to disable target size checks. This option can also be passed to individual `user.click` calls in the 2nd parameter.
```
Can we include the default value here?

gerardo-rodriguez replied:
Done https://github.com/cloudfour/pleasantest/pull/248/commits/c962fa3cb102f3a9fb44f20ff2590c35b35bd6bc

Paul-Hebert commented on README.md:
```diff
@@ -492,11 +503,13 @@ See the [`PleasantestUtils`](#utilities-api-pleasantestutils) documentation.
 Clicks an element, if the element is visible and the center of it is not covered by another element. If the center of the element is covered by another element, an error is thrown. This is a thin wrapper around Puppeteer's [`ElementHandle.click` method](https://pptr.dev/#?product=Puppeteer&version=v10.4.0&show=api-elementhandleclickoptions). The difference is that `PleasantestUser.click` checks t ...
 
-**Actionability checks**: It refuses to click elements that are not [**attached**](#attached) or not [**visible**](#visible). You can override the visibility check by passing `{ force: true }`.
+**Actionability checks**: It refuses to click elements that are not [**attached**](#attached) or not [**visible**](#visible) or which have too small of a [**target size**](#target-size). You can override the visibility and target size checks by passing `{ force: true }`.
```
Nit:

```suggestion
**Actionability checks**: It refuses to click elements that are not [**attached**](#attached), not [**visible**](#visible) or which have too small of a [**target size**](#target-size). You can override the visibility and target size checks by passing `{ force: true }`.
```

Paul-Hebert commented on src/user-util/index.ts:
```diff
@@ -45,6 +45,87 @@ ${el}`;
+  const size = typeof targetSize === 'number' ? targetSize : 44;
+
+  if (width < size || height < size) {
+    if (el instanceof HTMLInputElement && el.type === 'checkbox') {
```
Smart switching to the label size. :+1: Out of curiosity do we have/plan to add more general actionability checks for un-labeled inputs?

Paul-Hebert replied:
Should radio buttons get the same treatment?

calebeby replied:
Radio buttons should, yeah 👍 

Other kinds of inputs, I'm not sure when it makes sense to call .click on them, since usually you would call .type on them. I guess you could have an input type="submit" that maybe you would click? But it should probably be large enough to click easily. And for other kinds of (text/number) inputs, I think they should generally be large enough to click easily.

@gerardo-rodriguez I'm curious what you think as well

> do we have/plan to add more general actionability checks for un-labeled inputs?

We will probably add all the checks from here, and decide which ones make sense for which methods and which kinds of elements: https://playwright.dev/docs/actionability/

gerardo-rodriguez replied:
> @gerardo-rodriguez I'm curious what you think as well


I agree that we could also include:
- radio buttons
- input type="submit"

😉

Paul-Hebert replied:
Is it common for `input type="submit"` to be labeled? I haven't run into that. If not it seems like it should need to be large enough by itself.

gerardo-rodriguez replied:
> Is it common for input type="submit" to be labeled?

Can you elaborate, @Paul-Hebert, I'm not sure I understand. 🤔

gerardo-rodriguez replied:
@Paul-Hebert Oh, I think I understand now. You were asking if it's common for an `input type="submit"` to have a `<label>`, correct?

I don't know that I've seen that pattern. It seems it'd be the `input type="submit"` by itself, which I agree in this case, it should be large enough on its own.

gerardo-rodriguez replied:
I'm looking into this now. 🔍 👀

Paul-Hebert replied:
Yeah, exactly :+1:

I assumed that radio buttons/checkboxes got special treatment because they're tiny by default but can be activated by clicking their label.

calebeby replied:
Yeah I think that input type="submit" is visually represented as a button anyway so requiring it to be large enough (without checking for a label) seems reasonable to me

gerardo-rodriguez replied:
I have refactored this file. 👍

Paul-Hebert commented on src/user-util/index.ts:
```diff
@@ -45,6 +45,87 @@ ${el}`;
+    const targetSizeMsg =
+      typeof targetSize === 'number'
+        ? `Target size of element is smaller than ${size}px × ${size}px`
+        : 'Target size of element does not meet W3C recommendation of 44px × 44px: https://www.w3.org/WAI/WCAG21/Understanding/target-size.html';
```
Would it be worth DRYing up these repeated error messages to make them easier to maintain?

gerardo-rodriguez replied:
I'm looking into this now. 🔍 👀

gerardo-rodriguez replied:
Done https://github.com/cloudfour/pleasantest/pull/248/commits/38de7787ed5ed6456da7fe099608b8724b3e358a

Paul-Hebert reviewed (changes requested):
This is a fantastic feature. Great work @gerardo-rodriguez and @calebeby!

I left a few non-blocking nits.

I do have one potentially blocking piece of feedback: Should the checkbox logic also apply to radio buttons? Are there any other difficult-to-style clickable inputs where the label suffices?

gerardo-rodriguez commented on README.md:
```diff
@@ -324,6 +324,12 @@ Ensures that the element is visible to a user. Currently, the following checks a
 
+#### Target size
+
+Ensures that the clickable region of an element is large enough to easily be activated by users, even users with reduced fine motor skills or while using a touch input device.
```
@calebeby How would you feel quoting this first paragraph so we don't have to reinvent the wheel?

> https://www.w3.org/WAI/WCAG21/Understanding/target-size.html

Source: https://www.w3.org/WAI/WCAG21/Understanding/target-size.html

gerardo-rodriguez replied:
We could even link to that resource also.

gerardo-rodriguez commented on docs/errors/target-size.md:
```diff
@@ -0,0 +1,107 @@
+
+## Background/Intent
+
+This error is intended to encourage developers and designers to use target sizes that are large enough for users to easily click.
```
```suggestion
This error is intended to encourage developers and designers to use target sizes that are large enough for users to easily click or touch.
```

gerardo-rodriguez commented on docs/errors/target-size.md:
```diff
@@ -0,0 +1,107 @@
+# Error: Cannot click element that is too small
```
❤️ this document 😃

gerardo-rodriguez commented on docs/errors/target-size.md:
````diff
@@ -0,0 +1,107 @@
+
+Setting the `targetSize` option changes the minimum width/height (in px) used to check elements when `user.click` is called. The `targetSize` option can be passed in several places to control the scope of the change:
+
+**On an individual call**:
+
+```ts
+await user.click(button, { targetSize: 30 /* px */ });
+```
+
+**For a single test**:
+
+```ts
+test(
+  'test name',
+  withBrowser(
+    {
+      user: {
+        targetSize: 30 /* px */,
+      },
+    },
+    async ({ user }) => {
+      await user.click(something);
+    },
+  ),
+);
+```
+
+**For a test file**:
+
+```ts
+import { configureDefaults } from 'pleasantest';
+
+configureDefaults({
+  user: { targetSize: 50 /* px */ },
+});
+```
````
These examples are _super_ helpful, thanks, @calebeby! 🎉

gerardo-rodriguez commented on docs/errors/target-size.md:
```diff
@@ -0,0 +1,107 @@
+
+### Approach 1: Increasing the target size
+
+Much of the time, increasing the target size is the correct solution to the problem. By doing this, you are making your site more accessible to users. Usually, increasing `padding` or setting a `min-width`/`min-height` is the easiest way to ensure an element is large enough.
```
I was wondering if we might include helpful resources? Like this one: https://css-tricks.com/looking-at-wcag-2-5-5-for-better-target-sizes/

gerardo-rodriguez replied:
What do you think about this suggestion? 🙂 
```suggestion
Much of the time, increasing the target size is the correct solution to the problem. By doing this, you are creating a more inclusive user experience. Usually, increasing `padding` or setting a `min-width`/`min-height` is the easiest way to ensure an element's target size is large enough.
```

gerardo-rodriguez commented on src/user-util/index.ts:
```diff
@@ -45,6 +45,87 @@ ${el}`;
   }
 };
 
+// Why is this hardcoded?
+//   So that the snapshots do not fail when a new version is released and all the error messages change
+// Why is this not pointing to `main`?
+//   So that if the docs are moved around or renamed in the future, the links in previous PT versions still work
+// Does this need to be updated before every release?
+//   No, only when the docs are changed
+const docsVersion = 'v2.0.0';
```
@calebeby Thank you for the helpful comment! 👍

gerardo-rodriguez commented:
@calebeby As far as I can tell, all tests were passing before [this commit](https://github.com/cloudfour/pleasantest/pull/248/commits/7ccf880151be65aedbf2fd8ddce3658170ac4072). But I don't understand why the tests failed with a docs change (adding a comma). Is my understanding correct in that it seems the tests suddenly started failing?

calebeby commented:
@gerardo-rodriguez :thinking: yeah that is confusing that it started failing. I'll look into it and see if I can reproduce it locally

calebeby commented:
@gerardo-rodriguez the test failure is due to changes on `main` that were not yet merged into this branch. I'll merge that in and fix it

gerardo-rodriguez commented:
> @gerardo-rodriguez the test failure is due to changes on `main` that were not yet merged into this branch. I'll merge that in and fix it

Ah, thank you, @calebeby! 🎉

gerardo-rodriguez commented:
@calebeby I refactored the `src/user-util/index.ts` logic. The inline snapshots are now failing but I'm not sure what to do. I _think_ they need to get updated but I don't know how to update them within CI. 

When you have a moment, can you take a look and let me know how I should handle this? Thanks! 👍

calebeby commented:
@gerardo-rodriguez you can make the snapshots pass in CI by forcing the elements to have a fixed size (like adding width/height etc) so that differences in font rendering or browser defaults (maybe?) won't change the element's size between local + CI

gerardo-rodriguez commented:
> TODO: Add changeset and decide whether to leave this open instead of merging it right away since it is a breaking change
> 
> It may make sense to do a minor release with accessibility snapshots, followed by a major change with the CSS changes I am working on as well as this PR.

@calebeby I have addressed all PR feedback, but these items from you are still outstanding. I'm not sure what is actionable for me to do.

calebeby commented on src/user-util/index.ts:
```diff
@@ -45,6 +45,102 @@ ${el}`;
+
+  // Element and label was too small
+  throw error`${errorMessage}
+${getTargetSizeMessage({ el, targetSize })}
+${getDimensionsMessage(el)}
+${el}
```
@gerardo-rodriguez I'm noticing that these error messages have more in common than different so mind if I rearrange a little bit to call a single function with the element and the "suggestion text", where the function would construct the error message?

gerardo-rodriguez replied:
@calebeby Go for it! 👍

calebeby commented:
@gerardo-rodriguez I finished refactoring! I think this PR is all good to go, except that I want to hold off on merging until after I do a minor release, because this is a breaking change.
